File: src/main/java/org/jcodec/containers/mxf/model/UL.java
package org.jcodec.containers.mxf.model;
import org.jcodec.common.StringUtils;
import org.jcodec.platform.Platform;

import static org.jcodec.common.Preconditions.checkNotNull;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * An UL class that wraps UL bytes, introduced to implement custom comparison
 * rules
 * 
 * <p>
 * SMPTE 298-2009
 * </p>
 * 4.2 SMPTE-Administered Universal Label A fixed-length (16-byte) universal
 * label, defined by this standard and administered by SMPTE.
 * 
 * @author The JCodec project
 * 
 */
public class UL {
    private final byte[] bytes;

    public UL(byte[] bytes) {
        checkNotNull(bytes);
        this.bytes = bytes;
    }

    public static UL newULFromInts(int[] args) {
        byte[] bytes = new byte[args.length];
        for (int i = 0; i < args.length; i++) {
            bytes[i] = (byte) args[i];
        }

        return new UL(bytes);
    }
    
    public static UL newUL(String ul) {
        checkNotNull(ul);
        String[] split = StringUtils.splitS(ul, ".");
        byte b[] = new byte[split.length];
        for (int i = 0; i < split.length; i++) {
            int parseInt = Integer.parseInt(split[i], 16);
            b[i] = (byte) parseInt;
        }
        return new UL(b);
    }

    @Override
    public int hashCode() {
        return ((bytes[4] & 0xff) << 24) | ((bytes[5] & 0xff) << 16) | ((bytes[6] & 0xff) << 8) | (bytes[7] & 0xff);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof UL))
            return false;

        byte[] other = ((UL) obj).bytes;

        for (int i = 4; i < Math.min(bytes.length, other.length); i++)
            if (bytes[i] != other[i])
                return false;

        return true;
    }

    public boolean maskEquals(UL o, int mask) {
        if(o == null)
            return false;
        byte[] other = o.bytes;
        mask >>= 4;
        for (int i = 4; i < Math.min(bytes.length, other.length); i++, mask >>= 1)
            if ((mask & 0x1) == 1 && bytes[i] != other[i])
                return false;

        return true;
    }

    private final static char[] hex = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };

    @Override
    public String toString() {
        if (bytes.length == 0) return "";
        char[] str = new char[bytes.length * 3 - 1];
        int i = 0;
        int j = 0;
        for (i = 0; i < bytes.length - 1; i++) {
            str[j++] = hex[(bytes[i] >> 4) & 0xf];
            str[j++] = hex[bytes[i] & 0xf];
            str[j++] = '.';
        }
        str[j++] = hex[(bytes[i] >> 4) & 0xf];
        str[j++] = hex[bytes[i] & 0xf];
        return Platform.stringFromChars(str);
    }

    public int get(int i) {
        return bytes[i];
    }

    public static UL read(ByteBuffer _bb) {
        byte[] umid = new byte[16];
        _bb.get(umid);
        return new UL(umid);
    }
}

File: src/main/java/org/jcodec/containers/mxf/model/UL.kt
package org.jcodec.containers.mxf.model

import org.jcodec.common.Preconditions
import org.jcodec.common.StringUtils
import org.jcodec.common.and
import org.jcodec.common.shr
import org.jcodec.platform.Platform
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * An UL class that wraps UL bytes, introduced to implement custom comparison
 * rules
 *
 *
 *
 * SMPTE 298-2009
 *
 * 4.2 SMPTE-Administered Universal Label A fixed-length (16-byte) universal
 * label, defined by this standard and administered by SMPTE.
 *
 * @author The JCodec project
 */
class UL(private val bytes: ByteArray) {
    override fun hashCode(): Int {
        return bytes[4] and 0xff shl 24 or (bytes[5] and 0xff shl 16) or (bytes[6] and 0xff shl 8) or (bytes[7] and 0xff)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is UL) return false
        val other = other.bytes
        for (i in 4 until Math.min(bytes.size, other.size)) if (bytes[i] != other[i]) return false
        return true
    }

    fun maskEquals(o: UL?, mask: Int): Boolean {
        var mask = mask
        if (o == null) return false
        val other = o.bytes
        mask = mask shr 4
        var i = 4
        while (i < Math.min(bytes.size, other.size)) {
            if (mask and 0x1 == 1 && bytes[i] != other[i]) return false
            i++
            mask = mask shr 1
        }
        return true
    }

    override fun toString(): String {
        if (bytes.size == 0) return ""
        val str = CharArray(bytes.size * 3 - 1)
        var i = 0
        var j = 0
        i = 0
        while (i < bytes.size - 1) {
            str[j++] = hex[bytes[i] shr 4 and 0xf]
            str[j++] = hex[bytes[i] and 0xf]
            str[j++] = '.'
            i++
        }
        str[j++] = hex[bytes[i] shr 4 and 0xf]
        str[j++] = hex[bytes[i] and 0xf]
        return Platform.stringFromChars(str)
    }

    operator fun get(i: Int): Int {
        return bytes[i].toInt()
    }

    companion object {
        @JvmStatic
        fun newULFromInts(args: IntArray): UL {
            val bytes = ByteArray(args.size)
            for (i in args.indices) {
                bytes[i] = args[i].toByte()
            }
            return UL(bytes)
        }

        @JvmStatic
        fun newUL(ul: String): UL {
            Preconditions.checkNotNull(ul)
            val split = StringUtils.splitS(ul, ".")
            val b = ByteArray(split.size)
            for (i in split.indices) {
                val parseInt = split[i].toInt(16)
                b[i] = parseInt.toByte()
            }
            return UL(b)
        }

        private val hex = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')

        @JvmStatic
        fun read(_bb: ByteBuffer): UL {
            val umid = ByteArray(16)
            _bb[umid]
            return UL(umid)
        }
    }

}